owa/app/js/components/manage-autogeneration-option/ManageAutogenerationOption.jsx

```diff
@@ -0,0 +1,455 @@
+          manualEntry: false,
+          autoGen: false
+        },
+        {
+          idType: 'OpenMRS ID',
+          location: 'Laboratory',
+          sourceName: 'Generator for OpenMRS ID',
+          manualEntry: false,
+          autoGen: true
+        }
+      ],
+      identifierTypes: [
+        { id: 1, name: 'OpenMRS ID'},
+        { id: 2, name: 'OpenMRS Identification Number'},
+        { id: 3, name: 'Old Identification Number '}
+      ],
+      locations: [
+        { id: 1, name: 'Armani Hospital' },
+        { id: 2, name: 'Buruju Hospital' },
+        { id: 3, name: 'Ocean view Clinic' },
+        { id: 4, name: 'Indiana Out patient ward' },
+        { id: 5, name: 'Laboratory' }
+      ],
+      genOptionAutogenerateSources: [
+        { id: 1, name: 'Generator for OpenMRS ID' },
+        { id: 2, name: 'Generator for Something Else' }
+      ],
+      modal: false,
+      firstLevelCompleted: false,
+      currentGenOption: {
+        identifierType: '',
+        currentLocation: '',
+        autogenerateSource: '',
+        autoGenerationEnabled: false,
+        manualEntryEnabled: false
+      },
+      saveGenOptionStatus: 'new',
+      errors: {},
+      alertVisible: false
+    };
+    this.getGenDetails = this.getGenDetails.bind(this);
+    this.toggle = this.toggle.bind(this);
+    this.createGenOptions = this.createGenOptions.bind(this);
+    this.deleteGenOption = this.deleteGenOption.bind(this);
+    this.nextStage = this.nextStage.bind(this);
+    this.onDismiss = this.onDismiss.bind(this);
+    this.onChange = this.onChange.bind(this);
+    this.saveNewAutoGenOptions = this.saveNewAutoGenOptions.bind(this);
+    this.editGenOption = this.editGenOption.bind(this);
+    this.editCurrentGenOption = this.editCurrentGenOption.bind(this);
+  }
+
+  getGenDetails(id) {
+    const selectedGenOption = this.state.genOptions[id];
+    this.setState({
+      saveGenOptionStatus: 'edit',
+      firstLevelCompleted: true,
+      modal: true,
+      currentGenOption: {
+        id,
+        identifierType: selectedGenOption.idType,
+        currentLocation: selectedGenOption.location,
+        autogenerateSource: selectedGenOption.sourceName,
+        autoGenerationEnabled: selectedGenOption.autoGen,
+        manualEntryEnabled: selectedGenOption.manualEntry
+      },
+    });
+  }
+
+  editCurrentGenOption() {
+    const { genOptions, currentGenOption } = this.state;
+    const { id, identifierType, currentLocation, autogenerateSource,
+        autoGenerationEnabled, manualEntryEnabled } = currentGenOption;
+    const selectedGenOption = genOptions[id];
+    selectedGenOption.idType = identifierType;
+    selectedGenOption.location = currentLocation;
+    selectedGenOption.sourceName = autogenerateSource;
+    selectedGenOption.autoGen = autoGenerationEnabled;
+    selectedGenOption.manualEntry = manualEntryEnabled;
+    genOptions[id] = selectedGenOption; 
+    this.setState({
```

Comment:
You may want to use functional `setState` here since I see you're updating the state with multiple values from the array.

Reply:
I tried to see how the functional setState will work in this regards and I discovered that this will still work since I'm only changing a part of the state and not the whole state. Meanwhile this is just like a demo implementation to depict what the actual implementation will be like.